binder/src/main/java/io/grpc/binder/internal/Outbound.java

```diff
@@ -0,0 +1,493 @@
+    SUFFIX_SENT,
+    CLOSED,
+  }
+
+  /*
+   * Represents the state of data we've sent in binder transactions.
+   */
+  @GuardedBy("this")
+  private State outboundState = State.INITIAL; // Represents what we've delivered.
+
+  // ----------------------------------
+  // For reporting to StatsTraceContext.
+  /** Indicates we're ready to send the prefix. */
+  private boolean prefixReady;
+
+  @Nullable private InputStream firstMessage;
+
+  @Nullable private Queue<InputStream> messageQueue;
+
+  /**
+   * Indicates we have everything ready to send the suffix. This implies we have all outgoing
+   * messages, and any additional data which needs to be send after the last message. (e.g.
+   * trailers).
+   */
+  private boolean suffixReady;
+
+  /**
+   * The index of the next transaction we'll send, allowing the receiver to re-assemble out-of-order
+   * messages.
+   */
+  @GuardedBy("this")
+  private int transactionIndex;
+
+  // ----------------------------------
+  // For reporting to StatsTraceContext.
+  private int numDeliveredMessages;
+  private int messageSize;
+
+  private Outbound(BinderTransport transport, int callId, StatsTraceContext statsTraceContext) {
+    this.transport = transport;
+    this.callId = callId;
+    this.statsTraceContext = statsTraceContext;
+  }
+
+  final StatsTraceContext getStatsTraceContext() {
+    return statsTraceContext;
+  }
+
+  /** Call to add a message to be delivered. */
+  @GuardedBy("this")
+  final void addMessage(InputStream message) throws StatusException {
+    if (messageQueue != null) {
+      messageQueue.add(message);
+    } else if (firstMessage == null) {
+      firstMessage = message;
+    } else {
+      messageQueue = new ConcurrentLinkedQueue<>();
+      messageQueue.add(message);
+    }
+  }
+
+  @GuardedBy("this")
+  protected final void onPrefixReady() {
+    this.prefixReady = true;
+  }
+
+  @GuardedBy("this")
+  protected final void onSuffixReady() {
+    this.suffixReady = true;
+  }
+
+  // =====================
+  // Updates to delivery.
+  @GuardedBy("this")
+  private void onOutboundState(State outboundState) {
+    checkTransition(this.outboundState, outboundState);
+    this.outboundState = outboundState;
+  }
+
+  // ===================
+  // Internals.
+  @GuardedBy("this")
+  protected final boolean messageAvailable() {
+    if (messageQueue != null) {
+      return !messageQueue.isEmpty();
+    } else if (firstMessage != null) {
+      return numDeliveredMessages == 0;
+    } else {
+      return false;
+    }
+  }
+
+  @Nullable
+  @GuardedBy("this")
+  private final InputStream peekNextMessage() {
+    if (numDeliveredMessages == 0) {
+      return firstMessage;
+    } else if (messageQueue != null) {
+      return messageQueue.peek();
+    }
+    return null;
+  }
+
+  @GuardedBy("this")
+  private final boolean canSend() {
+    switch (outboundState) {
+      case INITIAL:
+        if (!prefixReady && !messageAvailable() && !suffixReady) {
```

Comment:
Thanks for the thorough review. 8)

This was vaguely intentional, but it's clearly confusing.

We'll mark a server-side outbound instance with prefixReady when the headers have been provided, but since headers aren't technically required by the ServerCall, (nor are any messages), the prefix might not be set.

This should be clearer if we set prefixReady when those things happen. I.e. cl/374896286. (If you prefer that, lgtm it and I'll merge here).